examples/gcf-pubsub-vm-delete-event-handler/main.py

```diff
@@ -0,0 +1,660 @@
+            'region': self.function_region,
+        }
+        resource = Resource(labels=resource_labels, type='cloud_function')
+        log_entry = {
+            'log_name': self.log_name,
+            'labels': {
+                'event_id': self.event_id,
+            },
+            'severity': self.severity(),
+            'resource': resource,
+        }
+        return log_entry
+
+
+class NoOp(StructuredLog):
+    MESSAGE = "{} No action taken (NO_OP)"
+
+
+class IgnoredEventSubtype(StructuredLog):
+    """Log for when pub/sub event subtype is ignored e.g. GCE_OPERATION_DONE"""
+    MESSAGE = "No action taken, event_type is not GCE_API_CALL for {}"
+    SEVERITY = Severity.DEBUG
+    RESULT = Result.OK
+    LEVEL = logging.DEBUG
+    DETAIL = Detail.IGNORED_EVENT
+
+
+class NoMatches(StructuredLog):
+    MESSAGE = "{} matches no DNS records (NO_MATCHES)"
+    SEVERITY = Severity.DEBUG
+    RESULT = Result.OK
+    LEVEL = logging.DEBUG
+    DETAIL = Detail.NO_MATCHES
+
+
+class LostRace(StructuredLog):
+    MESSAGE = "{} does not exist, likely lost race (LOST_RACE)"
+    SEVERITY = Severity.WARNING
+    RESULT = Result.NOT_PROCESSED
+    LEVEL = logging.WARNING
+    DETAIL = Detail.LOST_RACE
+
+
+class VmNoIp(StructuredLog):
+    MESSAGE = "{} has no IP address (VM_NO_IP)"
+    SEVERITY = Severity.INFO
+    RESULT = Result.OK
+    LEVEL = logging.INFO
+    DETAIL = Detail.VM_NO_IP
+
+
+class RecordSkipped(StructuredLog):
+    MESSAGE = "{} does not match DNS record {} ({})"
+    SEVERITY = Severity.DEBUG
+    RESULT = Result.OK
+    LEVEL = logging.DEBUG
+    DETAIL = Detail.RR_MISMATCH
+
+    def __init__(self, vm_uri: str, event_id: str, record: dict):
+        super().__init__(vm_uri, event_id)
+        self.record = record
+
+    def message(self):
+        return self.MESSAGE.format(self.vm_uri, self.record['name'],
+                                   self.DETAIL.name)
+
+
+class NotARecord(RecordSkipped):
+    DETAIL = Detail.RR_NOT_A_RECORD
+
+
+class NameMismatch(RecordSkipped):
+    DETAIL = Detail.RR_NAME_MISMATCH
+
+
+class IpMismatch(RecordSkipped):
+    DETAIL = Detail.RR_IP_MISMATCH
+
+
+class RecordDeleted(StructuredLog):
+    MESSAGE = "{} matches DNS record {} deleted (RR_DELETED)"
+    SEVERITY = Severity.NOTICE
+    RESULT = Result.OK
+    LEVEL = logging.INFO
+    DETAIL = Detail.RR_DELETED
+
+    def __init__(self, vm_uri: str, event_id: str, dns_project: str,
+                 dns_managed_zone: str, record: dict, response: dict):
+        """The structure of record is a dns#resourceRecordSet
+
+        See: https://cloud.google.com/dns/docs/reference/v1/resourceRecordSets
+        """
+        super().__init__(vm_uri, event_id)
+        self.dns_project = dns_project
+        self.dns_managed_zone = dns_managed_zone
+        self.record = record
+        self.response = response
+
+    def message(self):
+        return self.MESSAGE.format(self.vm_uri, self.record['name'])
```

Comment:
If we're overriding this `message` method,  might as well define Message in line here with an f string:
```suggestion
        return f"{self.vm_uri} matches DNS record {self.record['name']} deleted (RR_DELETED)"
```

Reply:
Ack, will refactor all of the StructuredLog classes to use inline message format strings.

Reply:
Decided to keep this as is because the new use of ABC requires MESSAGE to be overridden anyway.  In this case, I think it makes sense to override message() as well because there are additional format field for this subclass.

Reply:
Ah, this finally sunk in.  Changing this to get ride of MESSAGE entirely and just always override message().